src/wallet/asyncrpcoperation_sendmany.cpp

```diff
@@ -191,791 +149,514 @@ void AsyncRPCOperation_sendmany::main() {
-            HDSeed seed = pwalletMain->GetHDSeedForRPC();
-            ovk = ovkForShieldingFromTaddr(seed);
-        }
+    LogPrint("zrpc", "%s: fee: %s\n", getId(), FormatMoney(fee_));
 
-        // Set change address if we are using transparent funds
-        // TODO: Should we just use fromtaddr_ as the change address?
-        CReserveKey keyChange(pwalletMain);
-        if (isfromtaddr_) {
-            LOCK2(cs_main, pwalletMain->cs_wallet);
-
-            EnsureWalletIsUnlocked();
-            CPubKey vchPubKey;
-            bool ret = keyChange.GetReservedKey(vchPubKey);
-            if (!ret) {
-                // should never fail, as we just unlocked
-                throw JSONRPCError(
-                    RPC_WALLET_KEYPOOL_RAN_OUT,
-                    "Could not generate a taddr to use as a change address");
-            }
+    CReserveKey keyChange(pwalletMain);
+    uint256 ovk;
 
-            CTxDestination changeAddr = vchPubKey.GetID();
-            builder_.SendChangeTo(changeAddr);
-        }
+    auto getDefaultOVK = [&]() {
+        HDSeed seed = pwalletMain->GetHDSeedForRPC();
+        return ovkForShieldingFromTaddr(seed);
+    };
 
-        // Select Sapling notes
-        std::vector<SaplingOutPoint> ops;
-        std::vector<SaplingNote> notes;
-        CAmount sum = 0;
-        for (auto t : z_sapling_inputs_) {
-            ops.push_back(t.op);
-            notes.push_back(t.note);
-            sum += t.note.value();
-            if (sum >= txValues.targetAmount) {
-                break;
-            }
-        }
+    auto setTransparentChangeRecipient = [&]() {
+        LOCK2(cs_main, pwalletMain->cs_wallet);
 
-        // Fetch Sapling anchor and witnesses
-        uint256 anchor;
-        std::vector<std::optional<SaplingWitness>> witnesses;
-        {
-            LOCK2(cs_main, pwalletMain->cs_wallet);
-            pwalletMain->GetSaplingNoteWitnesses(ops, witnesses, anchor);
-        }
+        EnsureWalletIsUnlocked();
+        CPubKey vchPubKey;
+        bool ret = keyChange.GetReservedKey(vchPubKey);
+        if (!ret) {
+            // should never fail, as we just unlocked
+            throw JSONRPCError(
+                RPC_WALLET_KEYPOOL_RAN_OUT,
+                "Could not generate a taddr to use as a change address");
+        }
+
+        CTxDestination changeAddr = vchPubKey.GetID();
+        builder_.SendChangeTo(changeAddr);
+    };
+
+    // FIXME: it would be better to use the most recent shielded pool change
+    // address for the wallet's default unified address account, and the
+    // associated OVK
+    std::visit(match {
+        [&](const FromAnyTaddr& fromAny) {
+            ovk = getDefaultOVK();
+            setTransparentChangeRecipient();
+        },
+        [&](const PaymentAddress& addr) {
+            std::visit(match {
+                [&](const libzcash::SproutPaymentAddress& addr) {
+                    ovk = getDefaultOVK();
```

Comment:
Once we have UAs we just use the external OVK for the transparent address for the appropriate account, or the legacy account otherwise. z_sendmany was previously using the transparentOvkForShielding for sprout as well, I believe, so I just maintained that here.

Reply:
Previously we _couldn't_ send Sprout->Sapling via `z_sendmany`, so there wasn't any OVK used for Sprout (since Sprout itself has no OVK usage, and neither does transparent).

Reply:
Sprout keys don't have OVKs. This seems fine to me.